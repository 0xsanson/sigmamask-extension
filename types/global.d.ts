// In order for variables to be considered on the global scope they must be
// declared using var and not const or let, which is why this rule is disabled
/* eslint-disable no-var */

import { Client } from '@sentry/types';

declare class Platform {
  openTab: (opts: { url: string }) => void;

  closeCurrentWindow: () => void;
}

export declare global {
  var platform: Platform;

  namespace jest {
    interface Matchers<R> {
      toBeFulfilled(): Promise<R>;
      toNeverResolve(): Promise<R>;
    }
  }

  var sentry: Client;
}
