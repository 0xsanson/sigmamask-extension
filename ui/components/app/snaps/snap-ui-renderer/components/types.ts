import { Component, SnapId, UserInputEventTypes } from '@metamask/snaps-sdk';

export type UIComponentParams<T extends Component> = {
  element: T;
  elementKeyIndex: {
    value: number;
  };
  snapId: SnapId;
  parentForm: string;
  // @TODO: export ComponentState type
  state: Record<string, string | Record<string, string>>;
  handleEvent: ({
    eventType,
    componentName,
    parentForm,
    value,
  }: {
    eventType: UserInputEventTypes | 'stateChange';
    componentName?: string;
    parentForm?: string;
    value?: Record<string, string> | string;
  }) => void;
};

export type UiComponent<T extends Component> = (
  params: UIComponentParams<T>,
) => {
  element: string;
  props?: Record<string, unknown>;
  children?: UiComponent<any> | UiComponent<any>[] | string;
};
