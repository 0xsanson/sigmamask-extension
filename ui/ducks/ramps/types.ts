export interface AggregatorNetwork {
  active: boolean;
  chainId: number;
  chainName: string;
  nativeTokenSupported: boolean;
}
