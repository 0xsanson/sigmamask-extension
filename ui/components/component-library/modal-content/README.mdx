import { Story, Canvas, ArgsTable } from '@storybook/addon-docs';

import { ModalContent } from './modal-content';

# ModalContent

`ModalContent` is the container for the modal dialog's content

<Canvas>
  <Story id="components-componentlibrary-modalcontent--default-story" />
</Canvas>

## Props

The `ModalContent` accepts all props below as well as all [Box](/docs/components-ui-box--default-story#props) component props

<ArgsTable of={ModalContent} />

### Children

Use the `children` prop to render the content of `ModalContent`

<Canvas>
  <Story id="components-componentlibrary-modalcontent--children" />
</Canvas>

```jsx
import { ModalContent, Text } from '../../component-library';

<ModalContent>
  <ModalHeader />
  <Text>
    Lorem ipsum dolor sit amet consectetur adipisicing elit. Distinctio,
    reiciendis assumenda dolorum mollitia saepe, optio at aliquam molestias
    omnis quae corporis nesciunt natus, quas tempore ut ullam eaque fuga. Velit.
  </Text>
  <LoremIpsum />
  <LoremIpsum />
  <LoremIpsum />
  <LoremIpsum />
  <LoremIpsum />
</ModalContent>;
```

### Size

Currently the `ModalContent` supports a single size, this decision was made after we ran an audit on all modal sizes in the extension codebase and found that all modals could be made to fit the `ModalContentSize.Sm`(360px) size.

If you do require a larger modal size you can use the `modalDialogProps` to add a className to override the default size.

<Canvas>
  <Story id="components-componentlibrary-modalcontent--size" />
</Canvas>

```jsx
import { ModalContent, Text } from '../../component-library';

<ModalContent>
  <Text marginBottom={4}>
    ModalContentSize.Sm default and only size 360px max-width
  </Text>
</ModalContent>

<ModalContent {...args} modalDialogProps={{ style: { maxWidth: 800 } }}>
  <Text marginBottom={4}>
    Using modalDialogProps and adding a className setting a max width
    (max-width: 800px)
  </Text>
</ModalContent>
```

### Modal Dialog Ref

Use the `modalDialogRef` prop to pass a ref to the dialog element in the `ModalContent` component. This is primarily used with the `closeOnOutsideClick` prop on the `Modal` component. It allows the `Modal` to close when the user clicks outside of the `ModalContent` component.

<Canvas>
  <Story id="components-componentlibrary-modalcontent--modal-dialog-ref" />
</Canvas>

```jsx
import React, { useEffect, useRef, useState } from 'react';
import { ModalContent, Text } from '../../component-library';

const [show, setShow] = useState(false);
const modalDialogRef = useRef<HTMLDivElement>(null);

const handleClickOutside = (event: MouseEvent) => {
  if (
    modalDialogRef?.current &&
    !modalDialogRef.current.contains(event.target as Node)
  ) {
    setShow(false);
  }
};

useEffect(() => {
  document.addEventListener('mousedown', handleClickOutside);
  return () => {
    document.removeEventListener('mousedown', handleClickOutside);
  };
}, []);

<Button onClick={() => setShow(true)}>Show ModalContent</Button>
{show && (
  <ModalContent>
    Click outside of this ModalContent to close
  </ModalContent>
)}
```
