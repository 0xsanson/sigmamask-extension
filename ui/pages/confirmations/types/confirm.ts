import { ApprovalControllerState } from '@metamask/approval-controller';
import {
  SimulationData,
  TransactionType,
} from '@metamask/transaction-controller';

export type TypedSignDataV1Type = {
  name: string;
  value: string;
  type: string;
}[];

export type SecurityAlertResponse = {
  reason: string;
  features?: string[];
  result_type: string;
  providerRequestsCount?: Record<string, number>;
  securityAlertId?: string;
};

export type SignatureRequestType = {
  chainId?: string;
  id: string;
  msgParams?: {
    from: string;
    origin: string;
    data: string | TypedSignDataV1Type;
    version?: string;
  };
  type: TransactionType;
  custodyId?: string;
  securityAlertResponse?: SecurityAlertResponse;
};

export type TransactionRequestType = {
  chainId?: string;
  actionId: number;
  id: string;
  txParams?: {
    data: string;
    from: string;
    gas: string;
    maxFeePerGas: string;
    maxPriorityFeePerGas: string;
    to: string;
    value: string;
  };
  type: TransactionType;
  custodyId?: string;
  securityAlertResponse?: SecurityAlertResponse;
  origin: string;
  simulationData: SimulationData;
};

export type Confirmation = SignatureRequestType | TransactionRequestType;

export type ConfirmMetamaskState = {
  confirm: {
    currentConfirmation?: Confirmation;
    isScrollToBottomNeeded?: boolean;
  };
  metamask: {
    pendingApprovals: ApprovalControllerState['pendingApprovals'];
    approvalFlows: ApprovalControllerState['approvalFlows'];
  };
};
