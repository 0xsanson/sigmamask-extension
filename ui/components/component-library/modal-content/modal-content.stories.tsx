import React, { useEffect, useRef, useState } from 'react';
import { ComponentStory, ComponentMeta } from '@storybook/react';

import Box from '../../ui/box';

import { DISPLAY, BLOCK_SIZES } from '../../../helpers/constants/design-system';

import { BUTTON_VARIANT, Button, Text, ModalHeader } from '..';

import { ModalContent } from './modal-content';
import { ModalContentSize } from './modal-content.types';

import README from './README.mdx';

export default {
  title: 'Components/ComponentLibrary/ModalContent',
  component: ModalContent,
  parameters: {
    docs: {
      page: README,
    },
  },
  argTypes: {
    className: {
      control: 'text',
    },
    children: {
      control: 'text',
    },
    size: {
      control: 'select',
      options: Object.values(ModalContentSize).map((value) =>
        value.toLowerCase(),
      ),
    },
  },
  args: {
    children: 'Modal Content',
  },
} as ComponentMeta<typeof ModalContent>;

const LoremIpsum = () => (
  <Text marginBottom={4}>
    Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nullam euismod
    tortor vitae nisi blandit, eu aliquam nisl ultricies. Donec euismod
    scelerisque nisl, sit amet aliquet nunc. Donec euismod, nisl vitae
    consectetur aliquam, nunc nunc ultricies nunc, eget aliquam nisl nisl vitae
    nunc. Donec euismod, nisl vitae consectetur aliquam, nunc nunc ultricies
    nunc, eget aliquam nisl nisl vitae nunc. Donec euismod, nisl vitae
    consectetur aliquam, nunc nunc ultricies nunc, eget aliquam nisl nisl vitae
    nunc. Donec euismod, nisl vitae consectetur aliquam, nunc
  </Text>
);

const Template: ComponentStory<typeof ModalContent> = (args) => (
  <ModalContent {...args} />
);

export const DefaultStory = Template.bind({});
DefaultStory.storyName = 'Default';

export const Children: ComponentStory<typeof ModalContent> = (args) => (
  <ModalContent {...args}>
    <ModalHeader marginBottom={4}>Modal Header</ModalHeader>
    <Text marginBottom={4}>
      The ModalContent with ModalHeader and Text components as children
    </Text>
    <LoremIpsum />
    <LoremIpsum />
    <LoremIpsum />
    <LoremIpsum />
    <LoremIpsum />
  </ModalContent>
);

enum ModalContentSizeStoryOption {
  Sm = 'sm',
  ClassName = 'className',
}

export const Size: ComponentStory<typeof ModalContent> = (args) => {
  const [show, setShow] = useState({
    sm: false,
    className: false,
  });
  const handleOnClick = (size: ModalContentSizeStoryOption) => {
    setShow({ ...show, [size]: !show[size] });
  };

  return (
    <>
      <Box display={DISPLAY.FLEX} gap={4}>
        <Button
          variant={BUTTON_VARIANT.SECONDARY}
          onClick={() => handleOnClick(ModalContentSizeStoryOption.Sm)}
        >
          Show sm size
        </Button>
        <Button
          variant={BUTTON_VARIANT.SECONDARY}
          onClick={() => handleOnClick(ModalContentSizeStoryOption.ClassName)}
        >
          Show className
        </Button>
      </Box>
      {show.sm && (
        <ModalContent {...args}>
          <Text marginBottom={4}>
            ModalContentSize.Sm default and only size 360px max-width
          </Text>
          <Button onClick={() => setShow({ ...show, sm: false })}>Close</Button>
        </ModalContent>
      )}
      {show.className && (
        <ModalContent {...args} modalDialogProps={{ style: { maxWidth: 800 } }}>
          <Text marginBottom={4}>
            Using modalDialogProps and adding a className setting a max width
            (max-width: 800px)
          </Text>
          <Button onClick={() => setShow({ ...show, className: false })}>
            Close
          </Button>
        </ModalContent>
      )}
    </>
  );
};

export const ModalDialogRef: ComponentStory<typeof ModalContent> = (args) => {
  const [show, setShow] = useState(false);
  const modalDialogRef = useRef<HTMLDivElement>(null);
  const handleClickOutside = (event: MouseEvent) => {
    if (
      modalDialogRef?.current &&
      !modalDialogRef.current.contains(event.target as Node)
    ) {
      setShow(false);
    }
  };
  useEffect(() => {
    document.addEventListener('mousedown', handleClickOutside);
    return () => {
      document.removeEventListener('mousedown', handleClickOutside);
    };
  }, []);
  return (
    <>
      <Button onClick={() => setShow(true)}>Show ModalContent</Button>
      {show && (
        <ModalContent {...args}>
          Click outside of this ModalContent to close
        </ModalContent>
      )}
    </>
  );
};
