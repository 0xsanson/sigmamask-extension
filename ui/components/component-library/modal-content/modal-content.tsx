import React, { forwardRef, useRef, useContext, useEffect } from 'react';
import classnames from 'classnames';

import {
  BackgroundColor,
  BorderRadius,
  BLOCK_SIZES,
  DISPLAY,
  JustifyContent,
  AlignItems,
} from '../../../helpers/constants/design-system';

import Box from '../../ui/box/box';

import { ModalFocus, ModalContext } from '../modal';

import { ModalContentProps, ModalContentSize } from './modal-content.types';

export const ModalContent = forwardRef(
  (
    {
      className = '',
      children,
      size = ModalContentSize.Sm,
      width,
      modalDialogProps,
      ...props
    }: ModalContentProps,
    ref: React.Ref<HTMLElement>,
  ) => {
    const modalDialogRef = useRef<HTMLElement>(null);
    const {
      autoFocus,
      initialFocusRef,
      finalFocusRef,
      restoreFocus,
      isClosedOnEscapeKey,
      isClosedOnOutsideClick,
      onClose,
    } = useContext(ModalContext);
    const handleEscKey = (event: KeyboardEvent) => {
      if (isClosedOnEscapeKey && event.key === 'Escape') {
        onClose();
      }
    };

    const handleClickOutside = (event: MouseEvent) => {
      if (
        isClosedOnOutsideClick &&
        modalDialogRef?.current &&
        !modalDialogRef.current.contains(event.target as Node)
      ) {
        onClose();
      }
    };

    useEffect(() => {
      document.addEventListener('keydown', handleEscKey);
      document.addEventListener('mousedown', handleClickOutside);

      return () => {
        document.removeEventListener('keydown', handleEscKey);
        document.removeEventListener('mousedown', handleClickOutside);
      };
    }, []);
    return (
      <Box
        className={classnames('mm-modal-content', className)}
        ref={ref}
        display={DISPLAY.FLEX}
        width={BLOCK_SIZES.SCREEN}
        height={BLOCK_SIZES.SCREEN}
        justifyContent={JustifyContent.center}
        alignItems={AlignItems.flexStart}
        padding={4}
        {...props}
      >
        <ModalFocus
          autoFocus={autoFocus}
          initialFocusRef={initialFocusRef}
          finalFocusRef={finalFocusRef}
          restoreFocus={restoreFocus}
          modalDialogRef={modalDialogRef}
        >
          <Box
            className={classnames('mm-modal-content__dialog', {
              [`mm-modal-content__dialog--size-${size}`]:
                !modalDialogProps?.width,
            })}
            as="section"
            role="dialog"
            aria-modal="true"
            backgroundColor={BackgroundColor.backgroundDefault}
            borderRadius={BorderRadius.LG}
            width={width || BLOCK_SIZES.FULL}
            marginTop={12}
            marginBottom={12}
            padding={4}
            ref={modalDialogRef}
            {...modalDialogProps}
          >
            {children}
          </Box>
        </ModalFocus>
      </Box>
    );
  },
);
