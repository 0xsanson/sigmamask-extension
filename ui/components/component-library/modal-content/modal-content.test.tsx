/* eslint-disable jest/require-top-level-describe */
import { render } from '@testing-library/react';
import React from 'react';

import { ModalContent } from './modal-content';
import { ModalContentSize } from './modal-content.types';

describe('ModalContent', () => {
  it('should render with text inside the ModalContent', () => {
    const { getByText, getByTestId } = render(
      <ModalContent data-testid="modal-content">test</ModalContent>,
    );
    expect(getByText('test')).toBeDefined();
    expect(getByText('test')).toHaveClass('mm-modal-content__dialog');
    expect(getByTestId('modal-content')).toHaveClass('mm-modal-content');
  });
  it('should match snapshot', () => {
    const { container } = render(<ModalContent>test</ModalContent>);
    expect(container).toMatchSnapshot();
  });
  it('should render with and additional className', () => {
    const { getByText, getByTestId } = render(
      <ModalContent data-testid="test" className="test-class">
        test
      </ModalContent>,
    );
    expect(getByTestId('test')).toHaveClass('test-class');
  });
  it('should render with size sm', () => {
    const { getByText } = render(
      <>
        <ModalContent>default</ModalContent>
        <ModalContent size={ModalContentSize.Sm}>sm</ModalContent>
      </>,
    );
    expect(getByText('sm')).toHaveClass('mm-modal-content__dialog--size-sm');
    expect(getByText('default')).toHaveClass(
      'mm-modal-content__dialog--size-sm',
    );
  });
  it('should render with a ref', () => {
    const ref = React.createRef<HTMLDivElement>();
    render(<ModalContent modalDialogRef={ref}>test</ModalContent>);
    expect(ref.current).toBeDefined();
  });
  it('should render with additional props being passed to modalDialogProps', () => {
    const { getByTestId } = render(
      <ModalContent
        modalDialogProps={{ 'data-testid': 'test' }}
        data-testid="modal-content"
      >
        test
      </ModalContent>,
    );
    expect(getByTestId('test')).toBeDefined();
  });
});
