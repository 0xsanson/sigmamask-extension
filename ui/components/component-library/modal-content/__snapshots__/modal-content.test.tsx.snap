// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModalContent should match snapshot 1`] = `
<div>
  <section
    aria-modal="true"
    class="box mm-modal-content mm-modal-content--size-sm box--padding-4 box--flex-direction-row box--width-full box--background-color-background-default box--rounded-lg"
    role="dialog"
  >
    test
  </section>
</div>
`;
